### PR TITLE
Gather sort

### LINKED ISSUE #32: gather should returns a lexicographically sorted metric family slice

### PR DESCRIPTION
**Pull request by overvenus**
Resolved #32 

PTAL @siddontang @disksing

### CODE REVIEW

File: src/registry.rs [hunk 1]
```diff
@@ -66,7 +67,7 @@ impl RegistryCore {
     }
 
     fn gather(&self) -> Vec<proto::MetricFamily> {
-        let mut mf_by_name = HashMap::new();
+        let mut mf_by_name = HashMap::with_capacity(self.colloctors_by_id.len());
```
Comment (siddontang) [inline src/registry.rs]: you can use BtreeMap directly.

File: src/registry.rs [hunk 2]
```diff
@@ -90,8 +91,43 @@ impl RegistryCore {
...
+                    }
+                }
+
+                // We should never arrive here. Multiple metrics with the same
+                // label set in the same scrape will lead to undefined ingestion
+                // behavior. However, as above, we have to provide stable sorting
+                // here, even for inconsistent metrics. So sort equal metrics
+                // by their timestamp, with missing timestamps (implying "now")
+                // coming last.
+                m1.get_timestamp_ms().cmp(&m2.get_timestamp_ms())
```
Comment (siddontang) [inline src/registry.rs]: need to check timesamp_ms is none?

File: src/registry.rs [hunk 3]
```diff
@@ -221,4 +259,77 @@ mod tests {
...
+        map4.insert("cc", "12");
+        map4.insert("c1", "0");
+        map4.insert("a2", "da");
+        map4.insert("c0", "你好");
+        counter_vec.with(&map4).inc();
+        counter_vec.with(&map4).inc();
+        counter_vec.with(&map4).inc();
+        counter_vec.with(&map4).inc();
+
+        // # HELP test test help
```
Comment (siddontang) [inline src/registry.rs]: why comment this?

Comment (overvenus) [inline src/registry.rs]: That is the golang_client outputs.

Comment (siddontang) [inline src/registry.rs]: got it

File: src/registry.rs (hunk 2 above)
Comment (overvenus) [inline src/registry.rs]: Ported from client_golang, that is the way client_golang choose, rust should be the same, as it is ported from client_golang.

File: src/registry.rs [hunk 4]
```diff
@@ -221,4 +259,77 @@ mod tests {
...
+        map3.insert("c0", "hello");
+        counter_vec.with(&map3).inc();
+        counter_vec.with(&map3).inc();
+        counter_vec.with(&map3).inc();
+
+        let mut map4 = HashMap::new();
+        map4.insert("cc", "12");
+        map4.insert("c1", "0");
+        map4.insert("a2", "da");
+        map4.insert("c0", "你好");
```
Comment (siddontang) [inline src/registry.rs]: why Chinese?

Comment (overvenus) [inline src/registry.rs]: Label value supports Chinese, It is just a test anyway.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**